public/app/plugins/datasource/loki/backendResultTransformer.ts

```diff
@@ -0,0 +1,38 @@
+import { DataQueryRequest, DataQueryResponse, DataFrame, isDataFrame, FieldType } from '@grafana/data';
+import { partition } from 'lodash';
+import { LokiQuery, LokiQueryType } from './types';
+import { makeTableFrames } from './makeTableFrames';
+
+function isMetricFrame(frame: DataFrame): boolean {
+  return frame.fields.every((field) => field.type === FieldType.time || field.type === FieldType.number);
+}
+
+export function transformBackendResult(
+  response: DataQueryResponse,
+  request: DataQueryRequest<LokiQuery>
+): DataQueryResponse {
+  const { data, ...rest } = response;
+
+  // in the typescript type, data is an array of basically anything.
+  // we do know that they have to be dataframes, so we make a quick check,
+  // this way we can be sure, and also typescript is happy.
+  const dataFrames = data.map((d) => {
+    if (!isDataFrame(d)) {
+      throw new Error('transformation only supports dataframe responses');
+    }
+    return d;
+  });
+
+  const instantRefIds = new Set(
+    request.targets.filter((query) => query.queryType === LokiQueryType.Instant).map((query) => query.refId)
+  );
+
+  const [instantMetricFrames, normalFrames] = partition(dataFrames, (frame) => {
```

Comment:
nit: `instantMetricFrames` -> `instantAndMetricFrames` to make it a little but more clear that we have here frames from instant queries and frames with metrics.